proposals/0001-metadata-providers.md

```diff
@@ -0,0 +1,489 @@
+# Metadata Providers
+
+Authors: [Moritz Lang](https://github.com/slashmo), [Konrad 'ktoso' Malawski](https://github.com/ktoso)
+
+## Introduction
+
+While global metadata attributes may be manually set on a `LogHandler` level, there's currently no way of reliably providing contextual, automatically propagated, metadata when logging with swift-log.
+
+## Motivation
+
+To benefit from tools such as [Distributed Tracing](https://github.com/apple/swift-distributed-tracing) it is necessary for libraries to make use of the trace information.
+
+Most notably, loggers should participate in tracing by including some trace metadata (such as e.g. a `trace-id`) when logging, as it transparently enables developers to benefit from log correlation using those IDs.
+
+Today, the only supported way of providing such metadata is to pass them along to each log call explicitly:
+
+```swift
+logger.info("first this ...", metadata: ["trace-id": MyTracingLibrary.currentTraceID])
+logger.info("... now this", metadata: ["trace-id": MyTracingLibrary.currentTraceID])
+```
+
+This comes with a couple of downsides:
+
+### Error-prone and repetitive
+
+It's easy to forget passing this metadata to _all_ log statements, resulting in an inconsistent debugging experience as these log statements cannot be found using correlation IDs.
+
+The repetitiveness and verboseness of logging multiple metadata in-line quickly becomes annoying and vastly decreases the signal-to-noise ratio of Swift code trying to be a good citizen and making use of log correlation techniques such as distributed tracing.
```

Comment:
Similar point as the above as long as the libraries make sure to set the correct metadata on the logger and then pass it along it is not as error prone.

Reply:
This approach works well with more static metadata, but not with dynamically changing information like a span ID.